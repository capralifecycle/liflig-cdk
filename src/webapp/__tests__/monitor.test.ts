import { App, type CfnElement, Stack } from "aws-cdk-lib"
import { Match, Template } from "aws-cdk-lib/assertions"
import * as cw_actions from "aws-cdk-lib/aws-cloudwatch-actions"
import * as sns from "aws-cdk-lib/aws-sns"
import "jest-cdk-snapshot"
import { WebappMonitor } from ".."

test("App monitor default vaules snapshot test", () => {
  const app = new App()

  const appMonitorStack = new Stack(app, "AppMonitorStack")

  new WebappMonitor(appMonitorStack, "WebappMonitor", {
    webappDomainName: "webappdomain.example",
    appMonitorName: "local",
    telemetries: ["errors", "http", "performance"],
    defaultAlarmAction: new cw_actions.SnsAction(
      new sns.Topic(appMonitorStack, "Topic"),
    ),
  })

  expect(appMonitorStack).toMatchCdkSnapshot()
})

test("App monitor has custom events alarm properties", () => {
  const app = new App()

  const appMonitorStack = new Stack(app, "AppMonitorStack")

  const alarmActionTopic = new sns.Topic(appMonitorStack, "Topic")

  const alarmAction = new cw_actions.SnsAction(alarmActionTopic)

  const monitor = new WebappMonitor(appMonitorStack, "WebappMonitor", {
    webappDomainName: "webappdomain.example",
    appMonitorName: "AppMonitor",
    telemetries: ["http", "errors", "performance"],
    defaultAlarmAction: alarmAction,
  })

  monitor.configureEventsAlarm({
    enableOkAlarm: true,
  })

  monitor.addErrorsAlarm()

  const template = Template.fromStack(appMonitorStack)

  const snsTopicLogicalId = appMonitorStack.getLogicalId(
    alarmActionTopic.node.findChild("Resource") as CfnElement,
  )

  template.hasResourceProperties("AWS::CloudWatch::Alarm", {
    AlarmActions: [
      Match.objectEquals({
        Ref: snsTopicLogicalId,
      }),
    ],
    OKActions: [
      Match.objectEquals({
        Ref: snsTopicLogicalId,
      }),
    ],
    MetricName: "RumEventPayloadSize",
    Namespace: "AWS/RUM",
  })

  template.hasResourceProperties("AWS::CloudWatch::Alarm", {
    AlarmActions: [
      Match.objectEquals({
        Ref: snsTopicLogicalId,
      }),
    ],
    MetricName: "JsErrorCount",
    Namespace: "AWS/RUM",
  })

  // Since adding error alarm, alarm count is now 2
  template.resourceCountIs("AWS::CloudWatch::Alarm", 2)
})
