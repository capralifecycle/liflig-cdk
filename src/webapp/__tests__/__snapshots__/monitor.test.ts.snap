// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`App monitor default vaules snapshot test 1`] = `
Object {
  "Resources": Object {
    "TopicBFC7AF6E": Object {
      "Type": "AWS::SNS::Topic",
    },
    "WebappMonitorAppMonitor3AFF5E53": Object {
      "Properties": Object {
        "AppMonitorConfiguration": Object {
          "AllowCookies": true,
          "EnableXRay": false,
          "GuestRoleArn": Object {
            "Fn::GetAtt": Array [
              "WebappMonitorRumRoleE2A6B583",
              "Arn",
            ],
          },
          "IdentityPoolId": Object {
            "Ref": "WebappMonitorRumIdentityPool14A4031C",
          },
          "SessionSampleRate": 1,
          "Telemetries": Array [
            "errors",
            "http",
            "performance",
          ],
        },
        "CustomEvents": Object {
          "Status": "DISABLED",
        },
        "CwLogEnabled": false,
        "Domain": "webappdomain.example",
        "Name": "local",
      },
      "Type": "AWS::RUM::AppMonitor",
    },
    "WebappMonitorEventsAlarmFB4FE35F": Object {
      "Properties": Object {
        "AlarmActions": Array [
          Object {
            "Ref": "TopicBFC7AF6E",
          },
        ],
        "AlarmDescription": "A high number of events are being sent to CloudWatch RUM for app monitor 'local'. This may lead to high costs, so you'll likely want to investigate this.",
        "ComparisonOperator": "GreaterThanThreshold",
        "Dimensions": Array [
          Object {
            "Name": "application_name",
            "Value": "local",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "RumEventPayloadSize",
        "Namespace": "AWS/RUM",
        "Period": 60,
        "Statistic": "SampleCount",
        "Threshold": 1000,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "WebappMonitorRoleAttachment003A8509": Object {
      "Properties": Object {
        "IdentityPoolId": Object {
          "Ref": "WebappMonitorRumIdentityPool14A4031C",
        },
        "Roles": Object {
          "unauthenticated": Object {
            "Fn::GetAtt": Array [
              "WebappMonitorRumRoleE2A6B583",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::Cognito::IdentityPoolRoleAttachment",
    },
    "WebappMonitorRumIdentityPool14A4031C": Object {
      "Properties": Object {
        "AllowUnauthenticatedIdentities": true,
      },
      "Type": "AWS::Cognito::IdentityPool",
    },
    "WebappMonitorRumRoleE2A6B583": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": Object {
                "ForAnyValue:StringEquals": Object {
                  "cognito-identity.amazonaws.com:amr": "unauthenticated",
                },
                "StringEquals": Object {
                  "cognito-identity.amazonaws.com:aud": Object {
                    "Ref": "WebappMonitorRumIdentityPool14A4031C",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "Federated": "cognito-identity.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": "rum:PutRumEvents",
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":rum:",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        Object {
                          "Ref": "AWS::AccountId",
                        },
                        ":appmonitor/local",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "allow-rum",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
